[FLINK-21229] add confluent schema registry ssl support

qingdao81 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull requests adds ssl support to the confluent schema registry avro serializer/deserializer. 

## Brief change log

- `CachedSchemaCoderProvider` was extended to accept and delegate additional config properties to confluents `CachedSchemaRegistryClient`
- additional factory methods were added to `ConfluentRegistryAvroSerializationSchema` and `ConfluentRegistryAvroDeserializationSchema` which allow to pass a config map which is forwarded to `CachedSchemaCoderProvider`
- by definition all properties in this map with the prefix `schema.registry` are used to setup a `SslFactory` which is used in the http client for schema registry requests, see: [CachedSchemaRegistryClient](https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java#L180-L188)

## Verifying this change

This change added tests and can be verified as follows:

- Added unit tests for `ConfluentRegistryAvroSerializationSchema`, `ConfluentRegistryAvroDeserializationSchema` and `CachedSchemaCoderProvider`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

vidyaranya92 commented:
Hi, Is there any expected timeline for releasing this?

qingdao81 commented:
I fixed the issues in the unit tests - now I'm waiting for someone to review this PR.

Samuel-Fiddis commented:
Will this change allow for the basic auth credentials to be passed to the confluent schema registry? i.e. will this PR solve the problem detailed here: https://stackoverflow.com/questions/66096859/failed-to-connect-to-confluent-platform-schema-registry-apache-flink-sql-confl?newreg=e47e09684e404f4f822e417e74016268

qingdao81 commented:
Good question - I just had a closer look into the confluent serializer code. Confluent has their own implementation of a `RestService` which takes care about http requests to the schema registry. The changes I did so far will only make sure that the ssl properties are forwarded to the `SslFactory` which is used by the `RestService`. I can add another change that auth properties are forwarded to the `RestService` as well. I can not really tell you if this then solves the issue from the stackoverflow thread which you mentioned, but in theory it should.

But I also don't really know how that works in combination with the TableEnvironment. For the StreamingEnvironment it was not possible to provide any config properties (except the schema registry url) to the (de)serializer. But in your example you  already provide a bunch of properties.

Samuel-Fiddis commented:
I've updated and compiled the source code appending the following to the CachedSchemaCoderProvider class with the SchemaCoder get() method:

import io.confluent.kafka.schemaregistry.client.SchemaRegistryClientConfig;
import io.confluent.kafka.schemaregistry.client.security.basicauth.UserInfoCredentialProvider;

@Override
public SchemaCoder get() {
  RestService restService = new RestService(url);
  if(registryConfigs != null && Objects.equals(registryConfigs.get(SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE), "USER_INFO")) {
                  UserInfoCredentialProvider provider = new UserInfoCredentialProvider();
                  provider.configure(registryConfigs);
                  restService.setBasicAuthCredentialProvider(provider);
  }
}
....

But it doesn't look like the basic auth properties are passed from the Table API through to the registryConfigs map. When I've hard coded the properties into the registryConfigs everything works fine and the confluent schema registry works with basic auth access.

qingdao81 commented:
I added another commit which forwards the `registryConfig` to the `RestService`:
https://github.com/apache/flink/pull/15808/commits/070930ab00fecf6d07c01f3d02f4894dd8625a3a#diff-5f30e0f1719a976a878d0d07e9e9d47fac7f3a5d6daa7631349439c23f5320ffR72

The `RestService` uses a `BasicAuthCredentialProvider`:

```
public void configure(Map<String, ?> configs) {
        String basicCredentialsSource = (String)configs.get("basic.auth.credentials.source");
        String bearerCredentialsSource = (String)configs.get("bearer.auth.credentials.source");
        if (isNonEmpty(basicCredentialsSource) && isNonEmpty(bearerCredentialsSource)) {
            throw new ConfigException(String.format("Only one of '%s' and '%s' may be specified", "basic.auth.credentials.source", "bearer.auth.credentials.source"));
        } else {
            if (isNonEmpty(basicCredentialsSource)) {
                BasicAuthCredentialProvider basicAuthCredentialProvider = BasicAuthCredentialProviderFactory.getBasicAuthCredentialProvider(basicCredentialsSource, configs);
                this.setBasicAuthCredentialProvider(basicAuthCredentialProvider);
            } else if (isNonEmpty(bearerCredentialsSource)) {
                BearerAuthCredentialProvider bearerAuthCredentialProvider = BearerAuthCredentialProviderFactory.getBearerAuthCredentialProvider(bearerCredentialsSource, configs);
                this.setBearerAuthCredentialProvider(bearerAuthCredentialProvider);
            }

            String proxyHost = (String)configs.get("proxy.host");
            Object proxyPortVal = configs.get("proxy.port");
            Integer proxyPort = proxyPortVal instanceof String ? Integer.valueOf((String)proxyPortVal) : (Integer)proxyPortVal;
            if (isValidProxyConfig(proxyHost, proxyPort)) {
                this.setProxy(proxyHost, proxyPort);
            }

        }
    }
```

That should hopefully do the same as your change.

AHeise commented:
Hi @qingdao81 , thanks for your contribution. I have done a quick review round and it looks good already. Could you please resolve the compile issue (probably with your latest commit).

qingdao81 commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProvider.java:
```diff
@@ -37,21 +40,36 @@
+            return new CachedSchemaRegistryClient(url, identityMapCapacity);
+        }
+        RestService restService = new RestService(url);
+        restService.configure(registryConfigs);
```
Yes indeed - good point I will revert this change. Thanks!

qingdao81 commented:
@tao-xiao-sn You are right the properties are not forwarded to the SQL Factory. I will update this code as well.

twalthr commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java:
```diff
@@ -37,4 +37,40 @@ private RegistryAvroOptions() {}
+                    .withDescription("Password for SSL truststore");
+
+    public static final ConfigOption<String> SCHEMA_REGISTRY_BASIC_AUTH_CREDENTIALS_SOURCE =
+            ConfigOptions.key("schema-registry.basic.auth.credentials.source")
```
dots are for hierarchies not for word separation, rewrite to `schema-registry.basic-auth.credentials-source`

qingdao81 replied:
These properties are forwarded to the confluent `RestService` which expects a `.` notation.

[SchemaRegistryClientConfig](https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClientConfig.java#L25)

That is even true for `schema-registry`. I think for the 2 already existing properties `schema-registry.url` and `schema-registry.subject` this doesn't matter because they are treated in a different way. (they are parsed and then used for confluent api calls)

To my understanding for the additional properties it will only work if we forward for example `schema.registry.basic.auth.credentials.source` to the confluent api. If we use the grouping notation the properties will not match anymore and will be ignored.

If Flink should work with the dot and slash separation for hierarchies I had to transform the input back to pure dot notation at some point before delegating it to the confluent api.

At least this is my understanding but maybe I'm missing something?

twalthr commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java:
```diff
@@ -37,4 +37,40 @@ private RegistryAvroOptions() {}
+                    .withDescription("Basic auth credentials source for schema registry");
+
+    public static final ConfigOption<String> SCHEMA_REGISTRY_BASIC_AUTH_USER_INFO =
+            ConfigOptions.key("schema-registry.basic.auth.user.info")
```
rewrite to `schema-registry.basic-auth.user-info`

twalthr commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchemaTest.java:
```diff
@@ -0,0 +1,86 @@
+import static org.junit.Assert.assertTrue;
+
+/** Tests for {@link ConfluentRegistryAvroDeserializationSchema}. */
+public class ConfluentRegistryAvroDeserializationSchemaTest {
```
Tests should always be well chosen. I have feeling that this test is not very useful and can be dropped.

twalthr commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -84,10 +86,47 @@ private ConfluentRegistryAvroDeserializationSchema(
+     * @param schema schema of produced records
+     * @param url url of schema registry to connect
+     * @param registryConfigs map with additional schema registry configs (for example ssl
+     *     properties: 'schema.registry.ssl.keystore.location',
```
avoid code duplication and simplify `map with additional schema registry configs (for example for SSL properties).`

twalthr commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroSerializationSchemaTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.junit.Assert.assertTrue;
+
+/** Tests for {@link ConfluentRegistryAvroSerializationSchema}. */
+public class ConfluentRegistryAvroSerializationSchemaTest {
```
Remove as well?

twalthr commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactoryTest.java:
```diff
@@ -127,6 +139,66 @@ public void testMissingSubjectForSink() {
 
+    @Test
+    public void testDeserializationSchemaWithOptionalProperties() {
+        final AvroRowDataDeserializationSchema expectedDeser =
```
Reduce code duplication and move shared parts with the test above into a helper method.

tao-xiao-sn commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:
```diff
@@ -136,6 +160,31 @@ public String factoryIdentifier() {
+        configOptions.add(SCHEMA_REGISTRY_SSL_TRUSTSTORE_LOCATION);
+        configOptions.add(SCHEMA_REGISTRY_SSL_TRUSTSTORE_PASSWORD);
+        configOptions.add(SCHEMA_REGISTRY_BASIC_AUTH_CREDENTIALS_SOURCE);
+        configOptions.add(SCHEMA_REGISTRY_BASIC_AUTH_USER_INFO);
```
i dont think this is sufficient, for example schema registry supports bearer token too
https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClientConfig.java#L34

I think a better solution is to pass all properties that start with `schema-registry` prefix to schema registry client so that we dont need be coupled with the implementation of schema registry. this is similar to how table api pass kafka properties to kafka

twalthr replied:
+1 to this, we should not come up with our own properties because this means more maintenance overhead in the long-term when new properties are introduced. A prefix should be sufficient.

qingdao81 replied:
Yeah I was thinking about this as well but I wasn't sure how that works with the [RegistryAvroOptions](https://github.com/apache/flink/blob/master/flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java)

To my understanding I have to add one `ConfigOption` for each property?

twalthr replied:
You are right, this is a shortcoming in the current FormatFactory design. In dynamic table factories you have access to the plain `Map<String, String>` for excluding prefixes but not in format factories. We should fix this in a follow up issue. I will create a JIRA ticket for it.

Until then let's do hard coded properties but already in prefixed format. Then we can replace those explicit options with wildcard options in the future.

twalthr replied:
I created https://issues.apache.org/jira/browse/FLINK-22878. But this might take a while because it would affect many components in Flink to solve it properly.

twalthr commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProviderTest.java:
```diff
@@ -0,0 +1,72 @@
+        configs.put("schema.registry.ssl.truststore.password", keystorePassword);
+        SSLSocketFactory sslSocketFactory = getSslSocketFactory(configs);
+
+        assertFalse(sslSocketFactory == null);
```
is it possible to also check what has been set?

twalthr reviewed (commented):
Some feedback from my side as well.

twalthr commented:
@qingdao81 please let me know when you can update the PR. I'm happy to review it again. I think a lot of users could benefit from this.

qingdao81 commented:
@twalthr I was out the last couple of days but I should have time to continue tomorrow or latest on Wednesday.

qingdao81 commented:
@twalthr I updated the MR. I debugged the code and noticed that the `CachedSchemaRegistryClient` expects all properties in a dot notation and additional ssl related properties in the format: `schema.registry.ssl.xxx`
and auth related properties in the format without the `schema.registry` prefix: `basic.auth.user.info`

I still followed your suggestion to define config groups with the combination of dash / dot notation and transformed the Flink specific format to the Confluent specific format in `CachedSchemaCoderProvider#formatRegistryConfigs`.

I removed the unit tests which were not really meaningful and added a some tests for basic and bearer auth.

Happy to get your feedback!

twalthr reviewed (approved):
Thanks for the update @qingdao81. I forgot to delete my comment about the property format . If we simply forward the plain properties (without prefix) in the future, we can use the original keys with dot syntax. I will fix this while merging. Otherwise the changes look good to me. I will also investigate whether we can merge this to 1.13 as well to unblock users.

qingdao81 commented:
@twalthr Sounds good to me. Let me know in case you need any further support.

twalthr commented:
I have solved a couple of infrastructure problems in the meantime (support for prefix map syntax, support for fallback keys in the FactoryUtil). I pushed the final commits to this PR. The changes should be in a mergeable state now.

Airblader commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -84,14 +86,50 @@ private ConfluentRegistryAvroDeserializationSchema(
+    }
+
+    /**
+     * Creates {@link ConfluentRegistryAvroDeserializationSchema} that produces {@link
+     * GenericRecord} using provided reader schema and looks up writer schema in Confluent Schema
+     * Registry.
```
nit (and same below):

```suggestion
     * Creates {@link ConfluentRegistryAvroDeserializationSchema} that produces {@link
     * GenericRecord} using the provided reader schema and looks up the writer schema in the Confluent Schema
     * Registry.
```

Airblader commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -84,14 +86,50 @@ private ConfluentRegistryAvroDeserializationSchema(
+     *
+     * @param schema schema of produced records
+     * @param url URL of schema registry to connect
+     * @param identityMapCapacity maximum number of cached schema versions (default: 1000)
```
Why is the default documented here? How would I pass an `int` that uses the default?

Airblader commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -115,7 +153,45 @@ private ConfluentRegistryAvroDeserializationSchema(
+     *
+     * @param tClass class of record to be produced
+     * @param url URL of schema registry to connect
+     * @param identityMapCapacity maximum number of cached schema versions (default: 1000)
```
Same as above

Airblader reviewed (approved):
I just have a couple of nit comments on the docs, but LGTM overall and +1.

clouddra commented:
Are the new options applied to `DebeziumAvroFormatFactory` as well?

twalthr commented:
@clouddra no they are not :( Seems we forgot to propagate them. I will open a follow up issue.
